### DISCUSSION TITLE
Pull image from container registry using fine-grained access token

### DISCUSSION THREAD

**Discussion by mkniazuk123** [community/community#38467] [Repositories] [answered]
Hello, is it possible to pull image from ghcr.io using fine-grained personal access token? I've tried to use the token with read permissions but image pull fails with 403 Forbidden status.

**Comment by ossanna16** [comment]
Hi there @mkniazuk123 and welcome to our community! Thank you for asking a great question 🙂
While you are waiting for a reply to your question, here are a few things you can do:

To get started, introduce yourself in our [official introduction thread](https://github.com/orgs/community/discussions/23204)
To connect with other new users, check out this Discussion to [Meet Your GitHub Community Team](https://github.com/orgs/community/discussions/28776), further your GitHub knowledge by working through our [GitHub Skill 🆙 exercises](https://github.com/orgs/community/discussions/32121), and participate in our [weekly Fun Friday chats](https://github.com/orgs/community/discussions/30912)

**Comment by arbuckle** [comment, marked as answer]
![image](https://user-images.githubusercontent.com/917382/202886970-f4eae50d-1b3c-496b-ad2d-57e4ed1d0d80.png)

Was just looking into this myself. From the docs, fine-grained tokens are not supported.
https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry

**Reply by mkniazuk123** [reply]
Thank you @arbuckle

**Comment by burn2delete** [comment]
I am also having this issue, seems like over the past few days PAT (classic) is now broken for NPM registry also.

**Comment by rmehner** [comment]
Is this on the road map? It's one of our main use cases for tokens and I'd love to use a finer grained token approach for this.

**Comment by michaelmworthington** [comment]
Same here. We're looking to enforce fine-grained tokens so we can take advantage of the Enterprise management of the tokens. This will be a blocker to enforcing the removal of classic tokens.

**Comment by cledesma** [comment]
Fine-grained access token is also not working for me for accessing the Docker registry gchr.io :( This would be great once it works though

**Reply by sbkg0002** [reply]
Any update on this?

**Comment by matejm** [comment]
Does GitHub have any plans on fixing this? I think that access token with access only to packages in specific repository is one of the most basic security features they could offer. For example GitLab allows access token generation for individual repo and I think that is much cleaner solution for deployment tokens.

**Comment by sammcj** [comment]
Any update on this @Github?

It seems crazy that in order to publish container images to the Github registry you must use a classic token that has full write access to the repo.

**Comment by na-jakobs** [comment]
Any updates on this? 🙏

**Comment by danbopes** [comment]
So looking into this a bit more: I can understand why the devs didn't initially add package support, because packages themselves are tied to an account, or an organization, and not to a specific repository. In our cases, if a package is linked to a repo (And permissions are inherited), it should be fairly trivial to check permission from a fine-grained access token if they have access to read packages that are linked to that repository.

I implore github to add functionality for fine-grained access tokens, as security is important: I don't want to give PAT tokens to my cluster which have more access than what is needed.

**Comment by originswift-sys** [comment]
Okay, so "repository contents" in fine-grained tokens wasn't it, and I'm never getting back the time I wasted...
Got it 👍

**Comment by dummy-andra** [comment]
Hi guys, I was able to pull successfully from Github Registry using Fine-Grained token

Repository permissions
 Read access to actions, commit statuses, and metadata
 Read and Write access to code and pull requests
 
 
![image](https://github.com/community/community/assets/37038210/8471e29a-4ff6-4596-a1c7-44c4a6b1b3a1)
![image](https://github.com/community/community/assets/37038210/753dc957-8c13-4f1d-8248-5d7f261afb66)


Can you please try again on your end and let me know how it works?


Thanks,
Andra

**Reply by TBG-FR** [reply]
Sadly I get `Error response from daemon: denied` where the Personnal Access Token was working well 😞 

However the `docker login ghcr.io` command answered with `Login Succeeded` (tried with my username, and also the organization name)

**Reply by mkesper** [reply]
What pull requests  (PRs) have to do with pulling packages? I suspect you had somehow authenticated successfully to the registry before as this does NOT WORK AT ALL.

**Comment by mike667** [comment]
@dummy-andra Can it work for pushing? I tried to set read-write to all permissions, but actions can't push image to github registry. Got 403.

**Reply by dummy-andra** [reply]
I used classical token to push to github registry.

I had a project in a repo that used classical token with full permissions and the pipeline from here dit the push.

Other projects in diferite repos that only needed to pull from github registry,  were able to do this using fine-grained tokens using the permissions listed in the picture above.

However I even if it worked to pull I faced issues next on the line because "To use the REST API to manage GitHub Packages, you must authenticate using a personal access token (classic)." 

So...looks that for now we are stuck using the classic

**Comment by EdmundsEcho** [comment]
Thank you to everyone that has shared how to grant "package read permission" using the fine-grained access token.  With that in mind, I was reminded how the "classic" provides the precise access rights - ironic.  Until there is what I would call "direct support" for providing read access to the organization's packages using the fine-grained access token, I might recommend sticking with <s>coke</s> classic token.

**Comment by n3mawashi** [comment]
Has this worked for anyone recently? I'm trying to get a docker pull working with a fine-grain token but I am having no luck. This is a repo in an org setup.

**Comment by HarshanaJayasekara** [comment]
Respones #2

**Comment by singatias** [comment]
Any update on this? I refuse to use classic PAT in our organization for security purposes.

**Comment by ndrslmpk** [comment]
Any updates on this?

**Comment by pujux** [comment]
this is such an important feature, it's crazy there hasn't been any response from GitHub on this for 2 years

**Reply by singatias** [reply]
> this is such an important feature, it's crazy there hasn't been any response from GitHub on this for 2 years

I've decided to give up on GitHub and start self-hosting gitea for our own repositories, with daily backups on S3 storage at minimal cost. GitHub has too many security issues and a large attack surface. Feel free to copy my solution https://git.openharbor.io/Open-Harbor/gitea

**Reply by pujux** [reply]
> > this is such an important feature, it's crazy there hasn't been any response from GitHub on this for 2 years
> 
> 
> 
> I've decided to give up on GitHub and start self-hosting gitea for our own repositories, with daily backups on S3 storage at minimal cost. GitHub has too many security issues and a large attack surface. Feel free to copy my solution https://git.openharbor.io/Open-Harbor/gitea

Working on a large scale enterprise project so we're most definitely moving to either Harbor, DockerHub or Azure depending on what our DevOps team chooses to use. I like your alternative though! 🙌

It's just very disappointing this doesn't get any recognition by GitHub as it should be a rather trivial feature which many people and organizations would extremely benefit from...

**Reply by singatias** [reply]
> > > this is such an important feature, it's crazy there hasn't been any response from GitHub on this for 2 years
> > 
> > 
> > I've decided to give up on GitHub and start self-hosting gitea for our own repositories, with daily backups on S3 storage at minimal cost. GitHub has too many security issues and a large attack surface. Feel free to copy my solution https://git.openharbor.io/Open-Harbor/gitea
> 
> Working on a large scale enterprise project so we're most definitely moving to either Harbor, DockerHub or Azure depending on what our DevOps team chooses to use. I like your alternative though! 🙌
> 
> It's just very disappointing this doesn't get any recognition by GitHub as it should be a rather trivial feature which many people and organizations would extremely benefit from...

Nice I didn't know Harbor as a container registry, looks very interesting :)

**Comment by hjannasch** [comment]
I also need this feature. Anyone here from github who can place this request higher in the agenda? How long are fine-grained tokens in beta state now? Really annoying...

**Comment by yevgeniyo-ps** [comment]
This is trivial thing. We have restriction on org level to use classic tokens, from the other hand there is an ArgoCD that need to pull helm charts from ghcr....
what to do in such case? :-)

**Reply by pujux** [reply]
we have worked around the issue by creating a dummy github account that only has read access to this one repository and are using a token created from that user account 

it's ugly but the best option we have right now

**Reply by danbopes** [reply]
@pujux We ended up doing the same. So ugly, for what GitHub should have as a simple fix

**Comment by iloveicedgreentea** [comment]
I came here after trying to get this working for 10 minutes. Im actually shocked this isn't supported

**Reply by draganm** [reply]
same here. Such a basic feature is missing

**Reply by danbopes** [reply]
I've been watching, and they've been pushing it back release after release.

**Reply by hoppeeri** [reply]
i would not get my hopes up that this will ever be supported
https://jfrog.com/blog/github-jfrog-single-platform-experience/

**Comment by streetlogics** [comment]
I was able to finally get something resembling fine-grained tokens without having to create a fine-grained or classic personal access token. See the following links:
https://docs.github.com/en/packages/learn-github-packages/connecting-a-repository-to-a-package
https://docs.github.com/en/packages/learn-github-packages/about-permissions-for-github-packages#granular-permissions-for-userorganization-scoped-packages
https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token

**Reply by danbopes** [reply]
Github actions tokens have always been around, and allowed to push. The problem is: On the pull side. If you want to be able to pull from the remote repository (For example a deploy script), the `GITHUB_TOKEN` may work for the initial pull, but if you're on Kubernetes and the package gets deployed on another node: Good luck pulling. That github token is long expired, and useless.

I have not found ANY other way, other than to create another account, and join it to the repository as a collaborator. An utterly annoying workaround, but only way I could get it working securely.

**Comment by tommmlij** [comment]
Just wrapping my head around that.
With "Allow access via personal access tokens (classic) " disabled for my org, which is the sensible thing to do IMHO, I can push (did accomplish that in an action) but pull nowhere?
That makes the registry completely useless, or am I missing something?

**Comment by nyakaz73** [comment]
seriously GitHub, be serious on this, its it been 3 years now.

**Comment by whatwho** [comment]
It's been years now while this hasn't got supported! Come on Github, this is essential!

**Comment by MockingMagician** [comment]
Hi @ossanna16 any idea when, or even if it's planned?

**Reply by singatias** [reply]
At this point I assume the answer is "never"

**Reply by MockingMagician** [reply]
I'm desperate, if anyone recently usefull (listed below) can ping someone from Github coporation

@davevad93
@ishantmishra03
@HumayunK01
@linuxlawson
@Leorev01
@mte-tonmoy
@adityajethwa7
@Keshavkumar04
@AshishPandey26
@lucas-wall

**Reply by MockingMagician** [reply]
> At this point I assume the answer is "never"

I hope not ! :crying_cat_face:

**Reply by singatias** [reply]
I strongly suggest using dockerhub or an alternative at this point

**Comment by huweiATgithub** [comment]
This is insane, such a simple feature does not being implemented making the fine-grained tokens missing necessary features.
Besides, since it is possible to restrict access from classic token which is a good feature to enhance organization security.

But, ironically, once classic PAT is restricted, there is no way to access (write to public and read/write to private) container registry.

**Comment by thecarpetjasp** [comment]
Would be helpful if GitHub gave a bloody answer to this — which people have been asking and asking and asking for years!

Why should I be forced to use a PAT token which could leak every single repository of mine instead of using a fine grained token which I can scope to a single repo?

Having to create another account just to simply recreate this effect is such a gimmick. I appreciate GitHub a lot. But seriously? Just NO response at all!?

**Reply by hoppeeri** [reply]
Might be by design as GitHub have a partnership with JFrog:
https://github.blog/news-insights/product-news/github-and-jfrog/

Would not get my hopes up that this will ever come.

**Reply by danbopes** [reply]
Or all their pushes for AI integration everywhere...

**Comment by AvnanRahman** [comment]
Atleast add permission to:

write:packages
read:packages

This is what we need

**Comment by BasixKOR** [comment]
For folks who have been following along, GItHub has added this into their public roadmap: https://github.com/github/roadmap/issues/558

**Reply by danbopes** [reply]
It was added and removed several times, the latest being removed in 2024. It hasn't been readded since.

**Comment by Yiidiir** [comment]
@ankneis Do you have any insight into the process behind removing it from the roadmap? 
https://github.com/github/roadmap/issues/558

This is 2026 btw :)

**Comment by MockingMagician** [comment]
@github - Whenever you want guys ! Nobody is waiting for you :rofl: :exploding_head: :gun:

**Comment by zjpiazza** [comment]
god i hate github

**Reply by exasidn** [reply]
Still missing the feature

**Reply by NalinGHub** [reply]
not me wondering why tf the access token wasn't working 🤦‍♂️

**Comment by rinas21** [comment]
Problem

People want to pull container images from ghcr.io using fine-grained personal access tokens (PATs) instead of classic PATs.

Fine-grained PATs allow scoped, safer access, but GitHub currently doesn’t fully support them for container registry pulls.

Using a fine-grained token for read-only access often fails with 403 Forbidden, even if permissions seem correct.

Classic PATs work, but they are overly broad, giving access to all repos, which is a security risk.

Workarounds People Use

Classic PATs – still the most reliable way.

Separate GitHub account – create a dummy user, give it access to only the repo you need, generate a PAT, and use that. Ugly but works.

GitHub Actions GITHUB_TOKEN – works in workflows, but only for ephemeral pulls within that job, not for long-lived use like Kubernetes nodes.

Community Sentiment

Frustration over years of no support for fine-grained tokens for this use case.

Some speculate GitHub may never implement it due to other priorities or partnerships (e.g., JFrog).

The feature appeared on the public roadmap, was removed, and hasn’t returned.

Current Reality

For now, classic PATs are necessary to pull/push container images from GitHub Container Registry.

Fine-grained tokens cannot reliably replace classic PATs for registry access.

**Comment by MockingMagician** [comment]
As @rinas21 says, it must be due to the partnership or a reason we are not aware of, but why aren't they communicating about it? Just say yes or say no, we don't intend to do it. That will end the discussion. Please, @github, communicate about this!

**Comment by tuxillo** [comment]
they removed it from the public roadmap in 2024, doesn't seem they are willing to implement it 🫠

**Comment by pujux** [comment]
for anyone still searching for a solution, this is the only thing that reliably works: 

we have worked around the issue by creating a dummy github account that only has read access to this one repository and are using a token created from that user account

it's ugly but the best option we have until there is communication from github on this, which I don't think there will ever be

**Reply by obversary** [reply]
what ive been using as a work around recently is just using the actions pipeline to push image from container to ghcr using the internal token from github its not a Personal token but the internal  one ... i apologize if this has been answered already - you could in theory like you did just create a new token thats tethered to just packages but still a personalAT

**Reply by danbopes** [reply]
@architectfromthefuture Pushing to the repo has never really been the issue, as it's standard as part of the build process to deploy using CI_TOKEN. It's consuming the image via kubernetes or other process that a PAT token has a big security impact.

**Reply by obversary** [reply]
using actions pipeline is a diff authentication path hence not a PAT

**Reply by sdubrick** [reply]
@architectfromthefuture that's works for pushing to a registry, but like dan said that's not really the issue here. the problem here is controlling access for external services that need to pull the image securely. i think @pujux answer is the only way I've found to do it, or else just publish to a completely different container registry outside of github that has modern security controls

**Comment by zoikhash** [comment]
The enshitification of github continues

**Comment by HTop22** [comment]
### Docker login to GHCR with a Personal Access Token (classic)
Posting this method as found no method to use fine-grained access token. Pulling from a **private** GitHub Container Registry package requires authentication using `x-access-token` as the username, not your GitHub username.

### Setup
1. Create a **Personal Access Token (classic)** at `github.com/settings/tokens`
2. Select scope: `read:packages`
3. Login and pull:

```bash
echo "ghp_YOUR_TOKEN" | docker login ghcr.io -u x-access-token --password-stdin
docker pull ghcr.io/YOUR_ORG/YOUR_IMAGE:latest
```
Using `x-access-token` works reliably across personal accounts and organisation packages.

**Reply by danbopes** [reply]
Is this AI slop? This whole discussion is about not using PAT's, but being able to access packages via fine-grained access tokens.

**Reply by HTop22** [reply]
No method for fine-grained access tokens as per other forum comments.  Only method I found was to use `x-access-token` with Personal Access Token (classic) `read-packages`.  Working so in the absence of better solution thought it would be useful to post.

**Reply by danbopes** [reply]
The documentation already clearly explains how to connect to the docker registry with PAT [here](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-with-a-personal-access-token-classic)

**Reply by sdubrick** [reply]
>  Docker login to GHCR with a Personal Access Token (classic)

This issue is about how to avoid using classic tokens for registry access since they have security limitations compared to fine-grained tokens, such as not being able to manage and approve them at the organization level. Using classic tokens like this becomes a blocker for organizations needing to disable classic tokens completely.

**Comment by Kong-KEAT** [comment]
Hi, there. Can you guide me how to do it in the right way?

**Comment by hann-wang** [comment]
Our security team has disabled PAT (classic) in our organization and we lost access to all private images.

Need fine-grained tokens with container registry ASAP!

**Reply by obversary** [reply]
That sounds like the exact security deadlock that Github needs to clarify.

Your security team disabling classic tokens is a reasonable hardening move

If the only practical path for private image pulls is still classic PAT, then organizations are forced to choose between:
- keeping stronger token governance
or
- keeping private container workflows functional

That is such a bad tradeoff...

This is not just a convenience issue. Private image pulls are part of literally everything as well as production recovery paths...

Github docs need to clearly state whether fine-grained PATs can authenticate to ghcr.io for private image pull/push, what permissions are required, and what the recommended migration path is for organizations that have disabled classic PAT...

If the only working GHCR path requires classic PATs,then GitHub is effectively pressuring orgs to loosen a control they intentionally put in place...............

**Reply by singatias** [reply]
I strongly suggest to use dockerhub or an alternative for private images repository, github doesn't seems to want to fix this issue for years and now they have bigger fishes to catch.

**Reply by yakovliam** [reply]
> Our security team has disabled PAT (classic) in our organization and we lost access to all private images.
> 
> Need fine-grained tokens with container registry ASAP!

IMO, ideally, GitHub Apps should also be a first-class, fully supported mechanism for container registry workflows, so organizations can eliminate PAT-based access entirely.

Right now, the lack of clarity and parity between token types creates a security regression: teams are pushed toward weaker, legacy authentication simply to keep essential CI/CD and production systems running.

That tradeoff is not acceptable in a modern security model.

And @singatias, your short message has pushed me over the edge, and I will be pushing for a switch away from GitHub entirely.

**Reply by diofeher** [reply]
Please add this feature for Github Apps :( @github

**Comment by toby-griffiths** [comment]
Having an issue with this now, having just started for a new employer, and now I'm left with the challenge of accessing all things needed for work without having to expose my personal projects.  Can't believe that this is still an issue in today's world where we need to be so much more security conscious with the increasing threat levels!?

**Comment by robproject** [comment]
alright time for Forgejo
